src/quantile/mod.rs

```diff
@@ -541,57 +433,128 @@ where
+                    searched_indexes.push(lower_index(q, axis_len));
+                }
+                if I::needs_higher(q, axis_len) {
+                    searched_indexes.push(higher_index(q, axis_len));
+                }
+            }
+            searched_indexes.sort();
+            searched_indexes.dedup();
+
+            let mut results = Array::from_elem(results_shape, data.first().unwrap().clone());
+            Zip::from(results.lanes_mut(axis))
+                .and(data.lanes_mut(axis))
+                .apply(|mut results, mut data| {
+                    let index_map =
+                        get_many_from_sorted_mut_unchecked(&mut data, &searched_indexes);
+                    for (result, &q) in results.iter_mut().zip(qs) {
+                        let lower = if I::needs_lower(q, axis_len) {
+                            Some(index_map[&lower_index(q, axis_len)].clone())
+                        } else {
+                            None
+                        };
+                        let higher = if I::needs_higher(q, axis_len) {
+                            Some(index_map[&higher_index(q, axis_len)].clone())
+                        } else {
+                            None
+                        };
+                        *result = I::interpolate(lower, higher, q, axis_len);
+                    }
+                });
+            Ok(results)
+        }
+
+        quantiles_axis_mut(self.view_mut(), axis, qs.view(), interpolate)
+    }
+
+    fn quantile_axis_mut<I>(
+        &mut self,
+        axis: Axis,
+        q: N64,
+        interpolate: &I,
+    ) -> Result<Array<A, D::Smaller>, EmptyInput>
+    where
+        D: RemoveAxis,
+        A: Ord + Clone,
+        S: DataMut,
+        I: Interpolate<A>,
+    {
+        self.quantiles_axis_mut(axis, &aview1(&[q]), interpolate)
+            .map(|a| a.index_axis_move(axis, 0))
     }
 
-    fn quantile_axis_skipnan_mut<I>(&mut self, axis: Axis, q: f64) -> Array<A, D::Smaller>
+    fn quantile_axis_skipnan_mut<I>(
+        &mut self,
+        axis: Axis,
+        q: N64,
+        interpolate: &I,
+    ) -> Result<Array<A, D::Smaller>, EmptyInput>
     where
         D: RemoveAxis,
         A: MaybeNan,
         A::NotNan: Clone + Ord,
         S: DataMut,
         I: Interpolate<A::NotNan>,
     {
-        self.map_axis_mut(axis, |lane| {
+        if self.len_of(axis) == 0 {
+            return Err(EmptyInput);
+        }
+        let quantile = self.map_axis_mut(axis, |lane| {
             let mut not_nan = A::remove_nan_mut(lane);
             A::from_not_nan_opt(if not_nan.is_empty() {
                 None
             } else {
                 Some(
                     not_nan
-                        .quantile_axis_mut::<I>(Axis(0), q)
-                        .into_raw_vec()
-                        .remove(0),
+                        .quantile_axis_mut::<I>(Axis(0), q, interpolate)
+                        .unwrap()
```

Comment:
If we make invalid `q` an error instead of panic (see comment above), we need to update this. (The simplest solution would probably be to add a check at the top of the method body, like we're doing for `self.len_of(axis) == 0`, and keep the `unwrap` here.)